python/dolfin/io/xdmffile.py

```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2018 Chris N. Richardson, Garth N. Wells, Michal Habera
+#
+# This file is part of DOLFIN (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+import dolfin.cpp as cpp
+from dolfin.function.functionspace import FunctionSpace
+from dolfin.function.function import Function
+
+
+class XDMFFile(cpp.io.XDMFFile):
+
+    def read_checkpoint(self, V: FunctionSpace, name: str, counter: int=-1):
+        """Reads in function from checkpointing format
+
+        Parameters
+        ----------
+        V
+            Function space of saved function.
+        name
+            Name of function as saved into XDMF file.
+        counter : optional
+            Position of function in the file within functions of the same
+            name. Counter is used to read function saved as time series.
+            To get last saved function use counter=-1, or counter=-2 for
+            one before last, etc.
+
+        Note
+        ----
+        Parameter `V: Function space` must be the same as saved function space
+        except for ordering of mesh entities.
+
+        Returns
+        -------
+        dolfin.Function
+        """
+        if not isinstance(V, FunctionSpace):
```

Comment:
Could we allow `cpp.function.FunctionSpace` objects?

Reply:
Is there a strong need for that once rule "no `cpp.*` object being public" is applied?
I am afraid of mixing `cpp.*` (backend) with other objects. This will cause (and already caused in 2018.1) many confusion within users.

Reply:
We don't want any `cpp` objects being available without an explicit `dolfin.cpp` import. This doesn't mean a function can't accept a `cpp` object if, especially if handling it at the Python layer is tidier than handling it at the pybind11 layer.